R/process_sc_all_care_home.R

```diff
@@ -85,47 +96,63 @@ process_sc_all_care_home <- function(
       -"min_ch_provider",
       -"max_ch_provider"
     ) %>%
-    # tidy up ch_provider using 6 when disagreeing values
     tidyr::fill(.data[["ch_provider"]], .direction = "downup") %>%
     dplyr::ungroup()
 
 
-
   fixed_nursing_provision <- fixed_ch_provider %>%
     dplyr::group_by(
       .data[["sending_location"]],
       .data[["social_care_id"]],
       .data[["ch_admission_date"]]
     ) %>%
     # fill in nursing care provision when missing
-    # but present in the following entry
+    # but present in the following entry (n = 0)
     dplyr::mutate(
       nursing_care_provision = dplyr::na_if(.data[["nursing_care_provision"]], 9L)
     ) %>%
     tidyr::fill(.data[["nursing_care_provision"]], .direction = "downup")
 
 
   ready_to_merge <- fixed_nursing_provision %>%
-    # remove any duplicate records before merging for speed and simplicity
-    dplyr::distinct() %>%
+    # remove any duplicate records before merging
+    dplyr::distinct() %>% # (n = 3)
+    # sort data
+    dplyr::arrange(
+      .data[["sending_location"]],
+      .data[["social_care_id"]],
+      .data[["ch_admission_date"]],
+      .data[["period"]]
+    ) %>%
+    dplyr::group_by(
+      .data[["sending_location"]],
+      .data[["social_care_id"]],
+      .data[["ch_admission_date"]]
+    ) %>%
     # counter for split episodes
-    dplyr::mutate(
-      split_episode = tidyr::replace_na(
-        .data[["nursing_care_provision"]] != dplyr::lag(
-          .data[["nursing_care_provision"]]
-        ),
-        TRUE
-      ),
-      split_episode_counter = cumsum(.data[["split_episode"]])
+    # a split episode is an episode where the admission date is the same but the nursing provider has changed.
+    # We want to keep the nursing provision changes when we merge cases that have the same admission date
+    dplyr::mutate(previous_nursing_care_provision = dplyr::lag(.data[["nursing_care_provision"]])) %>%
+    # create a T/F flag for if nursing provision was the same as previous record with same admission date
+    dplyr::mutate(split_episode = tidyr::replace_na(.data[["previous_nursing_care_provision"]] != nursing_care_provision, TRUE)) %>%
+    dplyr::group_by(
+      .data[["social_care_id"]],
+      .data[["sending_location"]],
+      .data[["split_episode"]]
     ) %>%
-    dplyr::ungroup()
+    # create a count of each time the nursing provision changes between records with the same admission date
+    dplyr::mutate(split_episode_counter = ifelse(split_episode == TRUE, dplyr::row_number(), NA)) %>%
+    dplyr::group_by(
+      .data[["social_care_id"]],
+      .data[["sending_location"]]
+    ) %>%
+    # fill split episode counter. This will create a new id number for each different nursing provision within an episode
+    tidyr::fill(split_episode_counter, .direction = c("down")) %>%
+    dplyr::select(-previous_nursing_care_provision, -split_episode)
```

Comment:
Well done for dealing with this bit of code. It is very complex and took me a while to get my head around what is going on! I think the new logic is working well to split up episodes where they have changed due to a change in nursing provision or a new start date. I followed along with your example chi and could see where changes were happening. The fill `down` then resolves the counter and produces 3 continuous episodes in the particular example i was following. Very complex logic to work with!